File: imu_calib_python/imu_calib/calib/fit_gyro_g_sensitivity.py
from __future__ import annotations

import numpy as np

from imu_calib.models.data_structures import GsensRun
from imu_calib.utils.exceptions import ImuCalibError
from imu_calib.utils.math_utils import as_vector3, check_matrix_well_conditioned


def fit_gyro_g_sensitivity(gsens_runs: list[GsensRun], bg: np.ndarray, Cg: np.ndarray) -> tuple[np.ndarray, dict]:
    """Fit Gg from residuals using an explicit 9-parameter joint least-squares problem."""
    bg = as_vector3(bg, "bg")
    check_matrix_well_conditioned(Cg, "Cg")

    if not gsens_runs:
        return np.zeros((3, 3), dtype=float), {
            "message": "No gsens_runs provided. Returning zeros((3, 3)).",
            "valid": False,
            "gsens_term": "f_term",
            "gsens_definition": (
                "Gg is fit against gsens_runs.acc_ref and must be applied to the same "
                "f_term definition online."
            ),
            "num_runs": 0,
            "num_samples": 0,
            "design_rank": np.nan,
            "joint_design_rank": np.nan,
            "joint_design_condition": np.nan,
            "residual_rms": np.nan,
            "max_abs_residual": np.nan,
        }

    total_samples = 0
    for idx, run in enumerate(gsens_runs):
        _validate_run(run, idx)
        total_samples += run.gyro.shape[0]

    F = np.zeros((total_samples, 3), dtype=float)
    R = np.zeros((total_samples, 3), dtype=float)
    row_start = 0
    for run in gsens_runs:
        gyro = run.gyro
        acc_ref = run.acc_ref
        omega_ref = np.zeros_like(gyro) if run.omega_ref is None else run.omega_ref
        residual = (gyro - bg[None, :]) - (Cg @ omega_ref.T).T
        row_end = row_start + gyro.shape[0]
        F[row_start:row_end, :] = acc_ref
        R[row_start:row_end, :] = residual
        row_start = row_end

    design_rank = int(np.linalg.matrix_rank(F))
    if total_samples < 3 or design_rank < 3:
        return np.zeros((3, 3), dtype=float), {
            "message": "Insufficient gsens excitation. Returning zeros((3, 3)).",
            "valid": False,
            "gsens_term": "f_term",
            "gsens_definition": (
                "Gg is fit against gsens_runs.acc_ref and must be applied to the same "
                "f_term definition online."
            ),
            "num_runs": len(gsens_runs),
            "num_samples": total_samples,
            "design_rank": design_rank,
            "joint_design_rank": np.nan,
            "joint_design_condition": np.nan,
            "residual_rms": np.nan,
            "max_abs_residual": np.nan,
        }

    A = np.zeros((3 * total_samples, 9), dtype=float)
    y = np.zeros(3 * total_samples, dtype=float)
    for i in range(total_samples):
        rows = slice(3 * i, 3 * (i + 1))
        A[rows, :] = np.kron(F[i, :].reshape(1, 3), np.eye(3))
        y[rows] = R[i, :]

    joint_rank = int(np.linalg.matrix_rank(A))
    if joint_rank < 9:
        return np.zeros((3, 3), dtype=float), {
            "message": "Insufficient excitation for joint 9-parameter Gg fit. Returning zeros((3, 3)).",
            "valid": False,
            "gsens_term": "f_term",
            "gsens_definition": (
                "Gg is fit against gsens_runs.acc_ref and must be applied to the same "
                "f_term definition online."
            ),
            "num_runs": len(gsens_runs),
            "num_samples": total_samples,
            "design_rank": design_rank,
            "joint_design_rank": joint_rank,
            "joint_design_condition": np.inf,
            "residual_rms": np.nan,
            "max_abs_residual": np.nan,
        }

    x, *_ = np.linalg.lstsq(A, y, rcond=None)
    Gg = x.reshape(3, 3)
    fit_residual = y - A @ x
    return Gg, {
        "message": "Gg fitted using joint 9-parameter least squares.",
        "valid": True,
        "gsens_term": "f_term",
        "gsens_definition": (
            "Gg is fit against gsens_runs.acc_ref and must be applied to the same "
            "f_term definition online."
        ),
        "num_runs": len(gsens_runs),
        "num_samples": total_samples,
        "design_rank": design_rank,
        "joint_design_rank": joint_rank,
        "joint_design_condition": float(np.linalg.cond(A)),
        "residual_rms": float(np.sqrt(np.mean(fit_residual**2))),
        "max_abs_residual": float(np.max(np.abs(fit_residual))),
    }


def _validate_run(run: GsensRun, idx: int) -> None:
    if run.gyro.ndim != 2 or run.gyro.shape[1] != 3:
        raise ImuCalibError(f"gsens_runs[{idx}].gyro must have shape (N, 3).")
    if run.acc_ref.ndim != 2 or run.acc_ref.shape != run.gyro.shape:
        raise ImuCalibError(f"gsens_runs[{idx}].acc_ref must have the same shape as gyro.")
    if run.t.ndim != 1 or run.t.size != run.gyro.shape[0]:
        raise ImuCalibError(f"gsens_runs[{idx}].t length must match gyro samples.")
    if run.omega_ref is not None and run.omega_ref.shape != run.gyro.shape:
        raise ImuCalibError(f"gsens_runs[{idx}].omega_ref must have the same shape as gyro.")

File: imu_calib_matlab/src/calib/fit_gyro_g_sensitivity.m
function [Gg, info] = fit_gyro_g_sensitivity(gsensRuns, bg, Cg)
%FIT_GYRO_G_SENSITIVITY 用残差法拟合陀螺 g-灵敏度矩阵 Gg。
% 用法：
%   [Gg, info] = fit_gyro_g_sensitivity(gsensRuns, bg, Cg)
%
% 输入：
%   gsensRuns(i).gyro      : [N x 3] 陀螺原始测量
%   gsensRuns(i).acc_ref   : [N x 3] 用于拟合 Gg 的 specific force 输入项
%   gsensRuns(i).omega_ref : [N x 3] 可选参考角速度
%   bg                     : [3 x 1] 陀螺零偏
%   Cg                     : [3 x 3] 已估计的陀螺总矩阵
%
% 模型：
%   r = omega_m - bg - Cg * omega_ref
%   r ~= Gg * f
%
% 输出：
%   Gg   : [3 x 3] g-灵敏度矩阵
%   info : struct，包含有效性、设计矩阵秩、残差 RMS 等诊断信息
%
% 关键口径：
% - 本函数把 gsensRuns.acc_ref 视为 Gg 的输入项 f。
% - 因此在线补偿阶段也必须对同一种定义的 f 使用该 Gg，
%   否则会产生“拟合定义与补偿定义不一致”的误差。
% - 若数据不足，函数按工程约定返回 zeros(3) 并附带原因说明。

validateattributes(bg, {'numeric'}, {'vector', 'numel', 3, 'finite'}, mfilename, 'bg', 2);
validateattributes(Cg, {'numeric'}, {'size', [3, 3], 'finite'}, mfilename, 'Cg', 3);
bg = bg(:);

if nargin < 1 || isempty(gsensRuns) || numel(gsensRuns) == 0
    Gg = zeros(3);
    info = struct( ...
        'message', 'No gsensRuns provided. Returning zeros(3).', ...
        'valid', false, ...
        'gsens_term', 'sensor_axis_specific_force', ...
        'gsens_definition', ['Gg is fit against gsensRuns.acc_ref and must be applied to the ' ...
                             'same sensor-axis specific force definition online.'], ...
        'num_runs', 0, ...
        'num_samples', 0, ...
        'design_rank', NaN, ...
        'joint_design_rank', NaN, ...
        'joint_design_condition', NaN, ...
        'residual_rms', NaN, ...
        'max_abs_residual', NaN);
    return;
end

numRuns = numel(gsensRuns);
numSamplesPerRun = zeros(numRuns, 1);
for i = 1:numRuns
    validate_gsens_run(gsensRuns(i), i);
    numSamplesPerRun(i) = size(gsensRuns(i).gyro, 1);
end

totalSamples = sum(numSamplesPerRun);
F = zeros(totalSamples, 3);
R = zeros(totalSamples, 3);

rowStart = 1;
for i = 1:numRuns
    run = gsensRuns(i);
    gyro = run.gyro;
    accRef = run.acc_ref;

    if isfield(run, 'omega_ref') && ~isempty(run.omega_ref)
        omegaRef = run.omega_ref;
    else
        omegaRef = zeros(size(gyro));
    end

    % 先减去已知的 bg 和 Cg * omega_ref，剩余部分再尝试由 Gg * f 解释。
    modelTerm = (Cg * omegaRef.').';
    residual = bsxfun(@minus, gyro, bg.') - modelTerm;

    rowEnd = rowStart + numSamplesPerRun(i) - 1;
    F(rowStart:rowEnd, :) = accRef;
    R(rowStart:rowEnd, :) = residual;
    rowStart = rowEnd + 1;
end

if totalSamples < 3 || rank(F) < 3
    Gg = zeros(3);
    info = struct( ...
        'message', 'Insufficient gsens excitation. Returning zeros(3).', ...
        'valid', false, ...
        'gsens_term', 'sensor_axis_specific_force', ...
        'gsens_definition', ['Gg is fit against gsensRuns.acc_ref and must be applied to the ' ...
                             'same sensor-axis specific force definition online.'], ...
        'num_runs', numRuns, ...
        'num_samples', totalSamples, ...
        'design_rank', rank(F), ...
        'joint_design_rank', NaN, ...
        'joint_design_condition', NaN, ...
        'residual_rms', NaN, ...
        'max_abs_residual', NaN);
    return;
end

% 显式 9 参数联合最小二乘：
%   vec(r_i) = kron(f_i', I3) * vec(Gg)
% 这样能一次性求出完整的 3x3 Gg，而不是逐轴分别回归。
A = zeros(3 * totalSamples, 9);
y = zeros(3 * totalSamples, 1);
for i = 1:totalSamples
    rows = (3 * (i - 1) + 1):(3 * i);
    A(rows, :) = kron(F(i, :), eye(3));
    y(rows) = R(i, :).';
end

jointRank = rank(A);
if jointRank < 9
    Gg = zeros(3);
    info = struct( ...
        'message', 'Insufficient excitation for joint 9-parameter Gg fit. Returning zeros(3).', ...
        'valid', false, ...
        'gsens_term', 'sensor_axis_specific_force', ...
        'gsens_definition', ['Gg is fit against gsensRuns.acc_ref and must be applied to the ' ...
                             'same sensor-axis specific force definition online.'], ...
        'num_runs', numRuns, ...
        'num_samples', totalSamples, ...
        'design_rank', rank(F), ...
        'joint_design_rank', jointRank, ...
        'joint_design_condition', Inf, ...
        'residual_rms', NaN, ...
        'max_abs_residual', NaN);
    return;
end

x = A \ y;
Gg = reshape(x, 3, 3);

fitResidualVec = y - A * x;
fitResidual = reshape(fitResidualVec, 3, []).';

info = struct();
info.message = 'Gg fitted using joint 9-parameter least squares.';
info.valid = true;
info.gsens_term = 'sensor_axis_specific_force';
info.gsens_definition = ['Gg is fit against gsensRuns.acc_ref and must be applied to the ' ...
                         'same sensor-axis specific force definition online.'];
info.num_runs = numRuns;
info.num_samples = totalSamples;
info.design_rank = rank(F);
info.joint_design_rank = jointRank;
info.joint_design_condition = cond(A);
info.residual_rms = sqrt(mean(fitResidual(:) .^ 2));
info.max_abs_residual = max(abs(fitResidual(:)));
end

function validate_gsens_run(run, idx)
%VALIDATE_GSENS_RUN 检查单个 gsens run 的字段和维度是否合法。
requiredFields = {'gyro', 'acc_ref', 't'};
for i = 1:numel(requiredFields)
    if ~isfield(run, requiredFields{i})
        error('fit_gyro_g_sensitivity:MissingField', ...
            'gsensRuns(%d) is missing field "%s".', idx, requiredFields{i});
    end
end

validateattributes(run.gyro, {'numeric'}, {'2d', 'ncols', 3, 'nonempty', 'finite'}, ...
    mfilename, sprintf('gsensRuns(%d).gyro', idx));
validateattributes(run.acc_ref, {'numeric'}, {'2d', 'ncols', 3, 'nonempty', 'finite'}, ...
    mfilename, sprintf('gsensRuns(%d).acc_ref', idx));
validateattributes(run.t, {'numeric'}, {'column', 'nonempty', 'finite'}, ...
    mfilename, sprintf('gsensRuns(%d).t', idx));

if size(run.gyro, 1) ~= size(run.acc_ref, 1) || size(run.gyro, 1) ~= numel(run.t)
    error('fit_gyro_g_sensitivity:LengthMismatch', ...
        'gsensRuns(%d) fields gyro, acc_ref, and t must have the same length.', idx);
end

if isfield(run, 'omega_ref') && ~isempty(run.omega_ref) && size(run.omega_ref, 1) ~= size(run.gyro, 1)
    error('fit_gyro_g_sensitivity:OmegaRefLengthMismatch', ...
        'gsensRuns(%d).omega_ref length must match gyro length.', idx);
end
end
